### PR TITLE
respect user provided Boost prefix and throw an error if something g…

### PR DESCRIPTION
**Pull request by davydden**
…oes wrong

### CODE REVIEW

**Comment by davydden** [comment]
./run-tests

**Comment by tamiko** [comment]
I am not happy with this approach.

We do not error if a hint is ignored anywhere else, so why do it for boost?

File: cmake/configure/configure_boost.cmake [hunk 1]
```diff
@@ -87,6 +87,12 @@ ENDMACRO()
 MACRO(FEATURE_BOOST_FIND_EXTERNAL var)
   FIND_PACKAGE(BOOST)
 
+  IF(NOT "${BOOST_DIR}" STREQUAL "" AND NOT BOOST_FOUND)
+    MESSAGE(ERROR
+            "Could not find Boost in the provided prefix"
+           )
+  ENDIF()
```
Comment (tamiko) [inline cmake/configure/configure_boost.cmake]: This error message fires exactly if no external boost is found and `BOOST_DIR` is non-empty.

This is very misleading due to a couple of reasons. First, it does not imply that the external boost that is found is actually `BOOST_DIR`.

Secondly, this contradicts the current notion of `<FEATURE>_DIR` being a hint. So if we change that for boost, we have to change it for every feature (which will be a mess).

Third, if you do not want to pick up bundled libraries, simply set `DEAL_II_ALLOW_BUNDLED` to false.

**Comment by tjhei** [comment]
> We do not error if a hint is ignored anywhere else, so why do it for boost?

I guess it makes sense to treat boost, muparser, tbb, and umfpack (all our bundled libraries) similarly. On the other hand, boost is special because it is the only required one (right?).

**Comment by davydden** [comment]
@tamiko 

I am not happy about the current code which silently uses internally provided `boost` even when I set `-DBOOST_DIR` and other `cmake` variables to make sure external boost is found. That is not good and creates a mess at run-time when within a set of packages you can find all kind of combinations of `boost`. Say `trilinos` links against one `boost`, but `dealii` suddenly wants to use another.

I am happy to close this PR provided you implement an alternative:

Add a global flag which **prevents using any internally provided libraries under any circumstances**. If external boost is not found, I would like to have an error. If it's found, but is not functional -- throw an error. Same applies for other bundled libraries. I just did not have issues with them, but I already had issue with `boost`.

**Comment by Rombur** [comment]
> Add a global flag which prevents using any internally provided libraries under any circumstances. 

Isn't what `DEAL_II_ALLOW_BUNDLED` does?

**Comment by tamiko** [comment]
@davydden If you configure `-DDEAL_II_ALLOW_BUNDLED=false` the configure process will never consider any bundled libraries. In this case you get the following beautiful error message:
```
$ cmake -DDEAL_II_ALLOW_BUNDLED=FALSE ..
-- DEAL_II_WITH_BOOST has unmet external dependencies.
CMake Error at cmake/macros/macro_configure_feature.cmake:112 (MESSAGE):
  

  Could not find the boost library!

  Please ensure that a suitable boost library is installed on your computer.

  If the library is not at a default location, either provide some hints for
  autodetection,

      $ BOOST_DIR="..." cmake <...>
      $ cmake -DBOOST_DIR="..." <...>

  or set the relevant variables by hand in ccmake.

  Alternatively you may choose to compile the bundled library of ;boost by
  setting DEAL_II_ALLOW_BUNDLED=on or ;DEAL_II_FORCE_BUNDLED_BOOST=on.

Call Stack (most recent call first):
  cmake/macros/macro_configure_feature.cmake:268 (FEATURE_ERROR_MESSAGE)
  cmake/configure/configure_boost.cmake:148 (CONFIGURE_FEATURE)
  cmake/macros/macro_verbose_include.cmake:19 (INCLUDE)
  CMakeLists.txt:124 (VERBOSE_INCLUDE)
-- Configuring incomplete, errors occurred!
```

**Comment by davydden** [comment]
@tamiko @Rombur thanks for clarification. Apologies, it does work indeed exactly the way I wanted. I guess I just never noticed this option exists.

**Comment by tamiko** [comment]
@davydden Not, a problem. YW :-)

**Pull request closed without merging**